existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,295 @@
+# Existential and Universal Quantification in Haskell
+
+In this article I give a brief introduction into one of the most common Haskell extensions – `ExplicitForAll` – and syntax constructions that this extension brings to language.
+It is used in about all big haskell projects, so it is very nice to either become acquainted with it or to refresh it in your memory if you already know about it.
+
+Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints, and more!
+
+## Introduction
+
+By default, quantification in Haskell is implicit.
+Without the `ExplicitForAll` extension, you can't write quantifiers. Moreover, all variables are universally quantified, and existential quantification is off.
+The key point of both universal and existential quantification is choosing where, we are going to instantiate a type variable - where it is defined, or where it is used.
+
+## Universal quantification and `ExplicitForAll`
+
+By default, all type variables are considered universally quantified.
+This means that every time you use a function like `id`, you can _instantiate_ the universally quantified type variable `a` to (i.e., replace it with) any type.
+
+### Usage
+
+Let's try to add `forall` manually.
+The syntax is very simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
+the quantifier to introduce all type variables you want to use later:
+
+```Haskell
+func :: forall a b c m. a -> b -> m c -- OK, it compiles
+```
+Here we introduce four variables.
+All type variables that you introduce in your definition should belong to one `forall` exactly, though they may not belong to the same one. (I will explain it further.)
+
+```haskell
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+
+Let's add some constraints.
+For example, we might want to specify the `m` variable here as an instance of `Monad` typeclass.
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+```
+You might want to ask, what's so special about this syntax?
+Nothing changed since we added the quantifier because it was already there!
```

Comment:
It's confusing. If it's already there, why didn't `forall a b c . a -> b -> m c` work? `m` should be in the implicit, no?

Reply:
I don't get your point. We can't left `m` be quantified implicitly, or we get an error